### DISCUSSION TITLE
How to sum an Array or ArraySymbol over an axis?

### DISCUSSION THREAD

**Discussion by redeboer** [sympy/sympy#22269] [Q&A] [answered]
_Moving the ideas mentioned by @ThePauliPrinciple in https://github.com/sympy/sympy/pull/22265#issuecomment-942424065 to this separate discussion.  See also the indexing ideas mentioned by @ThePauliPrinciple in https://github.com/sympy/sympy/discussions/22219._

It seems that it is not yet possible to perform computations on the axes of (symbolic) arrays. This would be required for creating expressions that can be lambdified to `numpy.sum(..., axis=1)`.

To simplify things, let's stick with this sum as an example. We would like to do something like this:

```python
>>> from sympy.tensor.array.expressions.array_expressions import ArraySymbol
>>> A = ArraySymbol("A")
>>> expr = A.sum(axis=1)  # or some other syntax; probably requires a new Sum class
>>>
>>> from sympy.printing.numpy import NumPyPrinter
>>> NumPyPrinter().doprint(expr)
'numpy.sum(A, axis=1)'
>>>
>>> from sympy.printing.tensorflow import TensorflowPrinter
>>> TensorflowPrinter().doprint(expr)
'tf.math.reduce_sum(A, axis=1)'
```

Is this perhaps already possible? Or are there ideas for some implementation?

**Comment by ThePauliPrinciple** [comment]
Instead of a new Sum class `A.sum(axis=1)` could return an ArraySymbol which "remembers" that it still needs to sum over some index (but no longer allows accessing it), then the printer for ArraySymbol can take care of of printing the sum statement. I do like to point out though, that there would not be an easy translation for e.g. c code, where such a sum could require an assignment to a temporary variable.

**Reply by redeboer** [reply]
> Instead of a new Sum class `A.sum(axis=1)` could return an ArraySymbol which "remembers" that it still needs to sum over some index (but no longer allows accessing it)

I'd prefer to keep responsibilities separated, so putting this logic in another class. This will keep the logics simpler and more contained.

> I do like to point out though, that there would not be an easy translation for e.g. c code, where such a sum could require an assignment to a temporary variable.

Yes this is true, but if I understand correctly, that is what #22219 goes into (that is, working with code blocks). The example here would just  have to work for the conventional 'single-line' `lambdify`.

**Reply by ThePauliPrinciple** [reply]
A.sum(axis=1) is still potentially a SymbolicArray, I would expect to be able to e.g. do A.sum(axis=1).sum(axis=2). That's why I would rather see it as some marker being stored inside an ArraySymbol. Imagine the class ArrayIndexSum, which would behave similarly to your ArraySlice except it symbolically indicates a sum rather than a slice (correctly modifying the the shape value and __get__ behaviour).

**Reply by redeboer** [reply]
Hm okay fair point. Still, I would then not define a `.sum` method but just a class `ArraySum` that takes an `_ArrayExpr` and axes in its constructor. Same as `ArrayContraction` etc. This would be more scalable than a lot of if-else logic with additional attributes within a single class.

**Reply by ThePauliPrinciple** [reply]
The code printing of such a class could be something along the lines of:

```python
class ArraySum:
    def _tensorflowcode(expr,printer):
        return printer._module+'.reduce_sum({0},axis={1})'.format(printer._print(expr.args[0]),printer._print(expr.args[1:])))
```
if the first argument is the `_ArrayExpr` and further arguments are integers representing axes

**Reply by ThePauliPrinciple** [reply]
By using negative indices, we could also allow for arbitrary broadcasting of lamdify:
```python
class ArraySum:
    def _tensorflowcode(expr,printer):
        return printer._module+'.reduce_sum({0},axis={1})'.format(printer._print(expr.args[0]),printer._print([ i-len(expr.shape) for i in expr.args[1:]])))
```

**Reply by redeboer** [reply]
Aye, precisely :+1: 
My impression is that sympy mostly defines the print methods in the respective printer classes (here, through `TensorflowPrinter._print_ArraySum`), but the logics would be the same.

I'll wait for feedback to #22265 first – perhaps there are other ideas behind the design of array expressions. It seems that @Upabjojr wrote most of the array expression module.

**Reply by oscarbenjamin** [reply]
What's wrong with the existing Sum class?

**Reply by ThePauliPrinciple** [reply]
You are correct that this should be added to the TensorflowPrinter instead before merging, but this allows for testing new objects without integrating them into the library.

Would you consider ArraySlice to allow None (both numpy and tensorflow allow this as an alternative to newaxis), ArraySymbols and Arrays as valid slices? to replicate this behaviour:

![afbeelding](https://user-images.githubusercontent.com/35560762/137200133-5edd3483-47bf-4b30-ab1c-53f4f2830eb4.png)

**Reply by redeboer** [reply]
> What's wrong with the existing Sum class?

Do you mean for instance this?

```python
import sympy as sp
A = sp.ArraySymbol("A")
i = sp.Symbol("i")
sp.Sum(A[:, i], (i, 0, sp.oo))
```

Fine from a symbolic point of view, but can't lambdify to a vectorised form like `numpy.sum`. (In that sense, the discussion is perhaps too much oriented to code generation/lambdifying.)

**Reply by redeboer** [reply]
https://github.com/sympy/sympy/discussions/22269#discussioncomment-1472979
@ThePauliPrinciple thanks for the question, could you (re)post it in #22265?

**Reply by oscarbenjamin** [reply]
See also https://github.com/sympy/sympy/issues/5642#issuecomment-772789239

**Reply by ThePauliPrinciple** [reply]
I feel that sum could work:
```python
from sympy import Sum,Symbol,oo
i=Symbol('i')

class modNumPyPrinter(NumPyPrinter):
  def _print_Sum(printer,expr):
    print(expr)
    return ''

modNumPyPrinter().doprint(Sum(A[:,i,:],(i,0,oo)))
```

`_print_Sum` could be written to detect if there are any `ArraySlices` with the symbol over which it sums, then extract which axis this corresponds to. 

This does require a symbol to be a valid axis in ArraySlice (similar to an integer)

**Comment by redeboer** [comment]
> I think einsum can be correctly used for summing over the axis of a single Array. E.g.
> 
> ```python
> import numpy as np
> A=np.ones((3,))
> np.einsum('i->',A)
> ```
> 
> Would yield a scalar=3. I would imagine that it would shortcut to a special case and just run np.sum if it detects this is possible.
> 
> One could also imagine something like `i=Array.get_axis(1)` and then `Sum(Array,i)` could lead to printing np.sum(Array,axis=1).
> 
> If you need any help with the printing routines, or want me to give them an attempt that is no problem. I'm not very experienced with the Array part of sympy though, so I'm looking for your help on that side e.g giving an example of what sympy expression should lead to what print statement, or even simply the sympy expression (if it is clear what it should mean) that should be supported for printing.
> _https://github.com/sympy/sympy/pull/22265#issuecomment-942424065_

I'm not too familiar with [`einsum`](https://numpy.org/doc/stable/reference/generated/numpy.einsum.html), but it seems one can use it to sum over specific axes as well:

```python
>>> import numpy as np
>>> A = np.arange(9).reshape(3, 3); A
array([[0, 1, 2],
       [3, 4, 5],
       [6, 7, 8]])
>>>
>>> np.einsum("ij->i", A)  # np.sum(A, axis=1)
array([ 3, 12, 21])
>>> np.einsum("ji->i", A)  # np.sum(A, axis=0)
array([ 9, 12, 15])
>>>
>>> np.einsum("ij->i", A.transpose())
array([ 9, 12, 15])
>>> np.einsum(A, [0, 1], [0])  # np.sum(A, axis=1)
array([ 3, 12, 21])
```

The last one was added, because perhaps something can be done with a combination of `PermuteDims` (which lambdifies to `transpose`) and `ArrayContraction` (which lambdifies to `einsum`). Note that `ArrayContraction` lambdifies `einsum` with the index notation as shown in the end.

https://github.com/sympy/sympy/blob/3abd23db9fdefa4a7b2b745bd95d3d2f34311f3d/sympy/printing/numpy.py#L270-L274

**Reply by ThePauliPrinciple** [reply]
I think the problem with ArrayContraction is that it expects a pair of indices to be contracted on (which is the only sensible contraction for a Tensor, but not neccesarily for an Array and precludes summing over a single index). So I do not think that ArrayContraction will result in what you are looking for and that some new printing methods have to be written.

Note that with lambdify, it is expected that it can work with an arbitrary number of additional dimensions. 
Fortunately, this can be done with einsum since it accepts an ellipsis e.g. "...ij->...i".

**Reply by redeboer** [reply]
Ah ok thanks a lot. Yeah I'm not too familiar with tensor/array computations, so glad to read your thoughts on this.

At any rate, it seems clear that some class like `ArraySum` would have to be defined to get this to work.

**Reply by ThePauliPrinciple** [reply]
So currently, with my "Indexed story", I can obtain what you are looking for in the example:
![afbeelding](https://user-images.githubusercontent.com/35560762/137180000-36e069ac-a0ec-4c2e-a5f7-863bcfdc96f8.png)
But here it is the assignment class which takes care of the book keeping for the sum.

**Reply by Upabjojr** [reply]
> I think the problem with ArrayContraction is that it expects a pair of indices to be contracted on (which is the only sensible contraction for a Tensor, but not neccesarily for an Array and precludes summing over a single index)

We are not limited to contractions over two axes. That makes sense in differential geometry because you want to create invariants independent of the choice of the coordinate system.

However, there are other use cases. For example, in statistics, if you compute the gradient of a neural network, you may get a triple-index contraction.

**Reply by Upabjojr** [reply]
> Note that with lambdify, it is expected that it can work with an arbitrary number of additional dimensions.
> Fortunately, this can be done with einsum since it accepts an ellipsis e.g. "...ij->...i".

IIRC, in NumPy (but not in TensorFlow), `einsum` also accepts tuples of axis integer-valued indices.

**Reply by Upabjojr** [reply]
Another useful reference (may or may not be helpful): https://optimized-einsum.readthedocs.io/en/stable/#

**Comment by Upabjojr** [comment, marked as answer]
> It seems that it is not yet possible to perform computations on the axes of (symbolic) arrays. This would be required for creating expressions that can be lambdified to `numpy.sum(..., axis=1)`.

`ArrayContraction` can be given single axes:
```python
ArrayContraction(..., (1,))
```
This is equivalent to `numpy.sum(..., axis=1)`

**Reply by redeboer** [reply]
Okay sounds good :+1: 
So let's say we have the following expression.

```python
A = ArraySymbol("A")  # actually ArrayContraction requires a shape, is that really needed?
expr = ArrayContraction(A, (1,))
```

Should this then lambdify to `np.einsum(A, [0, 1, ...], [0, ...])`? And would this be equivalent to `np.sum(A, axis=1)`?

* If so, this would still have to be implemented, because `NumPyPrinter._print_ArrayContraction` currently can't handle this case.
* If so, does this mean there is no need for a class that can lambdify to `np.sum`? Then I will change this discussion into a question and mark this reply as the answer.

**Reply by Upabjojr** [reply]
`_print_ArrayContraction` can be modified. Maybe we need to disable the automatic normalization of array expressions first (i.e. if you nest two `ArrayContraction` calls, only one `ArrayContraction` will appear, maybe this should not be the default behaviour).

I would not recommend to create new classes concerning the array expressions module, unless it's really necessary. The main difficulty is that you should also find a way to normalize the array expression. Normalization algorithms are quite hard to write.

**Reply by redeboer** [reply]
Thanks! So I guess with #22279, this settles the discussion?

**Reply by Upabjojr** [reply]
Currently there is an auto-flattening mode (part of the normalization):
```
In [3]: A = ArraySymbol("A", 3, 3, 3, 3)

In [4]: ArrayContraction(ArrayContraction(A, (2,)), (0, 2))
Out[4]: ArrayContraction(A, (0, 3), (2,))
```

Maybe this should be disabled. For the code printer, it would be more convenient to separate multiple contractions from single contractions, and then handle the two nest `ArrayContraction` objects with different code printing logics. That is, reversing the normalization operation.

**Reply by ThePauliPrinciple** [reply]
I would argue the opposite, the code printer could always recursively call itself whenever multiple contractions are detected and only write one contraction (popping one contraction every time)  at a time if the language associated with that printer does not support summing over multiple indices in one call.

**Reply by Upabjojr** [reply]
Well, the idea is that you should be able to restructure the array expressions inside SymPy before passing it to the code printer. Currently this is not possible because `ArrayContraction` had automatic normalization by default and that cannot be disabled right now. So you will always have a standard contraction form.

It would make sense to apply some transformation rules before code printing, in order to create an optimized expression tree **inside** SymPy first, and then call the printer.

**Reply by ThePauliPrinciple** [reply]
Sorry, my point is only about the default behaviour, not to not have the option at all.

Would differently nested array contractions still lead to a mathematical identity in that case? (or zero when subtracted) As far as I understand, the order of contraction does not matter (although the index numbers might be different if the order is different). 

Of course, I understand that the order of contraction might matter (several orders of magnitude) when computing the contraction if it involves an array tensor product for the computational effort. 

Creating the optimized expression tree might be more appropriate in codegen.ast. The option to not have normalization would certainly be useful, but it feels to me that SymPy's default behaviour should be most convenient for the mathematical aspect and not the codegen aspect.

Reusing einsumpath from numpy might be helpful (requiring the user to give some approximation of the array sizes).

**Reply by Upabjojr** [reply]
> As far as I understand, the order of contraction does not matter (although the index numbers might be different if the order is different).

In a single `ArrayContraction` object, that is true and the index order does not matter. In case of nested contractions, the outer `ArrayContraction` object has less axes, so the contraction indices have to be renumbered (and trust me, renumbering the axes is not as simple as it sounds, though the algorithm is already implemented by now).

The point is, you may wish to have a NumPy code printer that either prints `np.einsum` or `np.sum` or both depending on whether the contractions are on multiple axes or on a single axis. The contraction axes have to be renumbered in either `np.einsum` or `np.sum`, depending on which on is the outer one.

> Creating the optimized expression tree might be more appropriate in codegen.ast.

Well, `sympy.tensor.array.expressions` is meant to be an AST for array expressions. More generally, all of SymPy is an AST for mathematical concepts.

> The option to not have normalization would certainly be useful, but it feels to me that SymPy's default behaviour should be most convenient for the mathematical aspect and not the codegen aspect.

I added default normalization to array expressions for simplicity reasons during the development. It should not be the default. There are three functions (`tensorcontraction`, `permutedims`, `tensordiagonal`) that should be used instead of calling the corresponding class constructors (`ArrayContraction`, `PermuteDims`, `ArrayDiagonal`). I haven't tested them extensively, so I'm not sure they always work, but my idea is that the functions should perform normalizations, while the class constructors shouldn't.

**Comment by Upabjojr** [comment]
For reference, I decided to base the class `ArrayContraction` on a similar [idea from Wolfram Mathematica](https://reference.wolfram.com/language/ref/TensorContract.html). Mathematica's API seems to me clearer and easier than NumPy's one in some cases.
